Invalid environment variable appear when using `deno install`ed command

deno 1.42.4 (release, x86_64-pc-windows-msvc)
v8 12.3.219.9
typescript 5.4.3

test.ts
```ts
import process from "node:process";
import { assertEquals } from "jsr:@std/assert"

const nodeKeys = [];
for (const key of Object.keys(process.env)) nodeKeys.push(key)
const denoKeys = [];
for (const key of Object.keys(Deno.env.toObject())) denoKeys.push(key)

nodeKeys.sort()
denoKeys.sort()

assertEquals(nodeKeys, denoKeys)
```

Output using `deno run -A test.ts`
```
Success!
```

Output using `deno install -frA -n testcmd test.ts` and `testcmd`
```
error: Uncaught (in promise) AssertionError: Values are not equal.


    [Diff] Actual / Expected


    [
+     "=C:",
      "ALLUSERSPROFILE",
      "AMDAPPSDKROOT",
```

Additionally, this new environment variable isn't getable

test.ts
```
for (const key of Object.keys(Deno.env.toObject())) 
    console.log(Deno.env.get(key))
```

Output using `deno install -frA -n testcmd test.ts` and `testcmd`
```
error: Uncaught (in promise) TypeError: Key contains invalid characters: "=C:"
    console.log(Deno.env.get(key))
                         ^
    at Object.getEnv [as get] (ext:runtime/30_os.js:102:10)
```

--- a/ext/os/lib.rs
+++ b/ext/os/lib.rs
@@ -223,10 +223,16 @@ fn op_env(
   state: &mut OpState,
 ) -> Result<HashMap<String, String>, PermissionCheckError> {
   fn map_kv(kv: (OsString, OsString)) -> Option<(String, String)> {
-    kv.0
-      .into_string()
-      .ok()
-      .and_then(|key| kv.1.into_string().ok().map(|value| (key, value)))
+    let key = kv.0.into_string().ok()?;
+    // Skip keys that `Deno.env.get`/`set`/`delete` would reject, so that
+    // enumeration stays consistent with the rest of the API. On Windows,
+    // cmd.exe exposes hidden per-drive cwd variables such as `=C:` which
+    // contain `=` and are not meant to be surfaced to users.
+    if key.is_empty() || key.contains(&['=', '\0'] as &[char]) {
+      return None;
+    }
+    let value = kv.1.into_string().ok()?;
+    Some((key, value))
   }
 
   let permissions_container = state.borrow::<PermissionsContainer>();

--- a/tests/unit/os_test.ts
+++ b/tests/unit/os_test.ts
@@ -48,6 +48,20 @@ Deno.test({ permissions: { env: true } }, function avoidEmptyNamedEnv() {
   assertThrows(() => Deno.env.delete("a\0a"), TypeError);
 });
 
+// Regression test for https://github.com/denoland/deno/issues/23443
+// Every key returned by `toObject()` must be one that `get()` accepts.
+// On Windows, cmd.exe injects hidden per-drive cwd variables such as `=C:`
+// that contain `=`, which used to be enumerated but were not gettable.
+Deno.test({ permissions: { env: true } }, function envToObjectKeysAreValid() {
+  for (const key of Object.keys(Deno.env.toObject())) {
+    assert(key.length > 0, "toObject() returned an empty key");
+    assert(!key.includes("="), `toObject() returned an invalid key: ${key}`);
+    assert(!key.includes("\0"), `toObject() returned an invalid key: ${key}`);
+    // Must not throw.
+    Deno.env.get(key);
+  }
+});
+
 Deno.test({ permissions: { env: false } }, function envPerm1() {
   assertThrows(() => {
     Deno.env.toObject();

